[JIT] Build SchemaInfo for understanding operator behavior

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80359
* #80244
* #80186
* __->__ #80177

- Created SchemaInfo class which publicizes op information through isMutating() and isAliasing() methods.
- Methods tested with sub_ op for mutation, aliasing, and exceptions in test_schema_info.cpp.

Differential Revision: [D37474862](https://our.internmc.facebook.com/intern/diff/D37474862)

Commit:
[JIT] Build SchemaInfo for understanding operator behavior

[ghstack-poisoned]

Commit:
Update on "[JIT] Build SchemaInfo for understanding operator behavior"


- Created SchemaInfo class which publicizes op information through isMutating() and isAliasing() methods.
- Methods tested with sub_ op for mutation, aliasing, and exceptions in test_schema_info.cpp.

[ghstack-poisoned]

Commit:
Update on "[JIT] Build SchemaInfo for understanding operator behavior"


- Created SchemaInfo class which publicizes op information through isMutating() and isAliasing() methods.
- Methods tested with sub_ op for mutation, aliasing, and exceptions in test_schema_info.cpp.

[ghstack-poisoned]

Commit:
Update on "[JIT] Build SchemaInfo for understanding operator behavior"


- Created SchemaInfo class which publicizes op information through isMutating() and isAliasing() methods.
- Methods tested with sub_ op for mutation, aliasing, and exceptions in test_schema_info.cpp.

[ghstack-poisoned]

Commit:
Update on "[JIT] Build SchemaInfo for understanding operator behavior"


- Created SchemaInfo class which publicizes op information through isMutating() and isAliasing() methods.
- Methods tested with sub_ op for mutation, aliasing, and exceptions in test_schema_info.cpp.

[ghstack-poisoned]

goldenxuett commented:
@goldenxuett has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37474862).

davidberard98 commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -0,0 +1,72 @@
+
+  if (lhsArg.alias_info() && lhsArg.alias_info()->isWildcardAfter() &&
+      rhsArg.alias_info() && rhsArg.alias_info()->isWildcardAfter()) {
+    if (lhsArg.type()->kind() == rhsArg.type()->kind()) {
```
https://github.com/pytorch/pytorch/blob/4ee29d6033f6ff4e1b75caa265d2646648b6636c/torch/csrc/jit/ir/alias_analysis.cpp#L86-L145 is a more complete description of the rules... not sure if this is copying the same behavior?

davidberard98 reviewed (approved):
Left some suggestions & some testing ideas

Also - we discussed offline but there's some internal changes you might need to make here in order to add test_schema_info.cpp into an internal TARGETS file.

Commit:
Update on "[JIT] Build SchemaInfo for understanding operator behavior"


- Created SchemaInfo class which publicizes op information through isMutating() and isAliasing() methods.
- Methods tested with sub_ op for mutation, aliasing, and exceptions in test_schema_info.cpp.

Differential Revision: [D37474862](https://our.internmc.facebook.com/intern/diff/D37474862)

[ghstack-poisoned]

goldenxuett commented:
@goldenxuett has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37474862).
